lib/spack/docs/packaging_guide.rst

```diff
@@ -612,6 +612,52 @@ it executable, then runs it with some arguments.
        installer = Executable(self.stage.archive_file)
        installer('--prefix=%s' % prefix, 'arg1', 'arg2', 'etc.')
 
+
+^^^^^^^^^^^^^^^^^^^^^^^^
+Deprecating old versions
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+There are many reasons to remove old versions of software:
+
+#. Security vulnerabilities (most serious reason)
+#. Changing build systems that increase package complexity
+#. Changing dependencies/patches/resources/flags that increase package complexity
+#. Maintainer/developer inability/unwillingness to support old versions
+#. No longer available for download (right to be forgotten)
+#. Package or version rename
+
+At the same time, there are many reasons to keep old versions of software:
+
+#. Reproducibility
+#. Requirements for older packages (e.g. some packages still rely on Qt 3)
+
+In general, you should not remove old versions from a ``package.py``. Instead,
+you should first deprecate them using the following syntax:
+
+.. code-block:: python
+
+   version('1.2.3', sha256='...', deprecated=True)
+
+
+This has two effects. First, ``spack info`` will no longer advertise that
+version. Second, commands like ``spack install`` that fetch the package will
+require user approval::
+
+   ==> Warning: This version is deprecated and may be removed in a future Spack release.
+   ==>   Fetch anyway? [y/N]
+
+
+If you use ``spack install --deprecated``, this check can be skipped.
+
+Version deprecations should last at least one full release cycle before the
```

Comment:
I REALLY REALLY like the pants policy on this: https://www.pantsbuild.org/docs/deprecation-policy. It also explicitly separates a deprecation policy for *pants options* (does spack have such a policy?) from the policy we require from plugin authors (equivalent to `package.py` maintainers):

> Plugin API deprecation policy
>
> When writing plugins, Pants is used as a library, rather than a binary. That is, you import Pants code to write plugins. We try to make this API stable for your plugins.

Reply:
If a deprecation is added in 0.16.0, I think we want to wait until after 0.17.0 to remove it. Users expect patch releases to only contain bug fixes, no functionality or hash changes. Other than that, I like your clarification.

Reply:
Mixed up semver there! Thanks.